### ISSUE TITLE
spark error due to missing cloud storage enum DURABLE_REDUCED_AVAILABILITY

### ISSUE THREAD

**Issue by lbergelson** [broadinstitute/gatk#2517]
We've made progress and the [Jenkins Tests](	at java.nio.file.Files.readAttributes(Files.java:1737) are failing with a new error now.  

```
com.google.cloud.storage.StorageClass.DURABLE_REDUCED_AVAILABILITY
	at org.broadinstitute.hellbender.engine.ReadsDataSource.<init>(ReadsDataSource.java:208)
	at org.broadinstitute.hellbender.engine.ReadsDataSource.<init>(ReadsDataSource.java:161)
	at org.broadinstitute.hellbender.engine.ReadsDataSource.<init>(ReadsDataSource.java:117)
	at org.broadinstitute.hellbender.engine.ReadsDataSource.<init>(ReadsDataSource.java:86)
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.getHeader(ReadsSparkSource.java:183)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeReads(GATKSparkTool.java:381)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeToolInputs(GATKSparkTool.java:361)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.runPipeline(GATKSparkTool.java:351)
	at org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram.doWork(SparkCommandLineProgram.java:38)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:116)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:179)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:198)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:121)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:142)
	at org.broadinstitute.hellbender.Main.main(Main.java:218)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:736)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:185)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:210)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:124)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.IllegalArgumentException: No enum constant com.google.cloud.storage.StorageClass.DURABLE_REDUCED_AVAILABILITY
	at java.lang.Enum.valueOf(Enum.java:238)
	at com.google.cloud.storage.StorageClass.valueOf(StorageClass.java:22)
	at com.google.cloud.storage.BlobInfo.fromPb(BlobInfo.java:940)
	at com.google.cloud.storage.Blob.fromPb(Blob.java:779)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:189)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:197)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.readAttributes(CloudStorageFileSystemProvider.java:571)
	at java.nio.file.Files.readAttributes(Files.java:1737)
	at java.nio.file.Files.isDirectory(Files.java:2192)
	at htsjdk.samtools.util.IOUtil.assertFileIsReadable(IOUtil.java:346)
	at org.broadinstitute.hellbender.engine.ReadsDataSource.<init>(ReadsDataSource.java:205)
	... 23 more
```

**Comment by lbergelson** [comment]
@jean-philippe-martin Could you investigate this one?

**Comment by jean-philippe-martin** [comment]
Sounds like a version mismatch. I'm on #2516 now but I can put this bug into my list. I'm going to need more background, though. Crucially I need to know how to reproduce this bug.

**Comment by lbergelson** [comment]
Hmn, I haven't reproduced it manually myself.  It's happening on the nightly jenkins tests that run on a google cluster.  I'm assuming it reproduces if you run the same command that their running on any dataproc cluster:

something along the lines of :
```
gatk-launch MarkDuplicatesSpark --shardedOutput true -O output.bam --numReducers 0 --apiKey $API_KEY -I gs://broad-gatk-test-jenkins/CEUTrio.HiSeq.WEx.b37.NA12892.readnamesort.bam -- ----sparkLauncher GCS --cluster $YOUR_CLUSTER
```

**Comment by lbergelson** [comment]
@jean-philippe-martin I think it's a mismatch between the server api and the google-cloud-java.

There are two different things that are both called storage class:  See this one: https://www.dartdocs.org/documentation/google_cloud_storage/0.4.1-pre2/file_storage/StorageClass-class.html

vs the java one: https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-storage/src/main/java/com/google/cloud/storage/StorageClass.java

It looks like these are orthogonal concepts, that happened to have the same name and an overlapping set of values.  I think they're getting mixed up.

**Comment by jean-philippe-martin** [comment]
Actually the page at https://cloud.google.com/storage/docs/storage-classes
(linked from the Java code) mentions all of the values in the code, plus `Durable Reduced Availability`; it says not to use the latter. 

It looks like all these terms apply to the same concept, though perhaps some values are deprecated and that's why they are not in the enum.

**Comment by lbergelson** [comment]
Oh, you're right, it's not in the big table so I didn't notice it.

**Comment by jean-philippe-martin** [comment]
I can confirm I was able to reproduce with:
```
./gatk-launch MarkDuplicatesSpark --shardedOutput true -O output.bam --numReducers 0 --apiKey $GOOGLE_API_KEY -I gs://broad-gatk-test-jenkins/CEUTrio.HiSeq.WEx.b37.NA12892.readnamesort.bam -- --sparkRunner GCS --cluster $MYCLUSTER
```

**Comment by jean-philippe-martin** [comment]
Interestingly, just adding the constant to gcloud allows gatk to proceed.

Well it crashed for me a bit later:

[Stage 0:==========================================>            (431 + 2) / 553]17/03/30 00:30:53 WARN org.apache.spark.scheduler.TaskSetManager: Lost task 431.0 in stage 0.0 (TID 431, jp-test-cluster-w-0.c.genomics-pipelines.internal): com.google.cloud.storage.StorageException: 503 Service Unavailable
Service Unavailable
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:186)
...
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.readAttributes(CloudStorageFileSystemProvider.java:571)
	at java.nio.file.Files.readAttributes(Files.java:1737)
	at java.nio.file.Files.isRegularFile(Files.java:2229)
	at htsjdk.samtools.SamFiles.lookForIndex(SamFiles.java:72)

That's the same 503 we've been protecting against in reads, now rearing its head on a readAttributes call.

**Comment by jean-philippe-martin** [comment]
To be fair I must point out that this test was running without retry settings (because the API for retries changed and I was just trying to test the attributes).

**Comment by jean-philippe-martin** [comment]
Fixed in upstream, so we should see this problem go away as soon as gcloud releases a new version and we switch to it.

**Comment by lbergelson** [comment]
👍

**Comment by jean-philippe-martin** [comment]
I reran this with retry enabled. It says it finished successfully. There was an error in the middle but it continued anyways (this isn't shown in the final output, but there was a second bar).
I'm not sure about the output, though. Where is this command supposed to leave `output.bam`? It's not on my desktop.

```
[Stage 1:=====================================>                 (375 + 2) / 553]17/03/30 18:18:46 WARN org.apache.hadoop.hdfs.DFSClient: DFSOutputStream ResponseProcessor exception  for block BP-369249695-10.240.0.8-1490738675068:blk_1073745922_5098
java.io.EOFException: Premature EOF: no length prefix available
	at org.apache.hadoop.hdfs.protocolPB.PBHelper.vintPrefixed(PBHelper.java:2282)
	at org.apache.hadoop.hdfs.protocol.datatransfer.PipelineAck.readFields(PipelineAck.java:244)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer$ResponseProcessor.run(DFSOutputStream.java:733)
17/03/30 18:18:46 WARN org.apache.hadoop.hdfs.DFSClient: Error Recovery for block BP-369249695-10.240.0.8-1490738675068:blk_1073745922_5098 in pipeline DatanodeInfoWithStorage[10.240.0.4:50010,DS-5596b1b5-b89c-4c39-bbd8-7423614eae0e,DISK], DatanodeInfoWithStorage[10.240.0.3:50010,DS-a0c20806-0af3-4679-b8cd-9cae6ca25071,DISK]: bad datanode DatanodeInfoWithStorage[10.240.0.4:50010,DS-5596b1b5-b89c-4c39-bbd8-7423614eae0e,DISK]
17/03/30 20:00:21 INFO org.spark_project.jetty.server.ServerConnector: Stopped ServerConnector@61cda923{HTTP/1.1}{0.0.0.0:4040}
20:00:21.366 INFO  MarkDuplicatesSpark - Shutting down engine
[March 30, 2017 8:00:21 PM UTC] org.broadinstitute.hellbender.tools.spark.transforms.markduplicates.MarkDuplicatesSpark done. Elapsed time: 165.11 minutes.
Runtime.totalMemory()=1222115328
Job [ac3f4131-f19f-47db-8cc3-82b243ad4b72] finished successfully.
```

**Comment by jean-philippe-martin** [comment]
Re-running this with a gs:// output seems to have worked perfectly! I now have a bunch of files there.

**Comment by jean-philippe-martin** [comment]
This should be fixed once #2581 is in; this is my reminder to try it then.

**Comment by lbergelson** [comment]
@jean-philippe-martin You should have the abiilty to manually trigger builds on jenkins if you sign in as well.  That would be an easy way to test if you don't want to wait for the nightly to run.

**Comment by jean-philippe-martin** [comment]
Are you saying it's OK to wait until #2581 is checked in and then change this PR to point to a nightly build and merge?

**Comment by jean-philippe-martin** [comment]
That's ok, it can wait until tomorrow. I'm also rerunning manually on my cluster to check - it looks like it's working perfectly so far!

**Comment by jean-philippe-martin** [comment]
This now worked fine when I ran it manually. Looks like the fix worked.

**Comment by jean-philippe-martin** [comment]
@lbergelson can you confirm that the nightly run passed, or alternatively tell me where to look?

**Comment by lbergelson** [comment]
@jean-philippe-martin Sorry for the slow response, I was away last week.  Nightlies are fixed.  Looks good!  fixed by #2581

**Comment by jean-philippe-martin** [comment]
great, thank you @lbergelson !
